Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,686 @@
+            # fallback. This'll mostly happen for "default" and "default_factory" where the
+            # type may be a C-type that can't be converted to Python.
+            return self._make_string()
+
+    def _make_string(self):
+        from .AutoDocTransforms import AnnotationWriter
+        writer = AnnotationWriter(description="Dataclass field")
+        string = writer.write(self.arg)
+        return ExprNodes.StringNode(self.pos, value=EncodedString(string))
+
+    def generate_evaluation_code(self, code):
+        return self.arg.generate_evaluation_code(code)
+
+
+def _set_up_dataclass_fields(node, fields, dataclass_module):
+    # For defaults and default_factories containing things like lambda,
+    # they're already declared in the class scope, and it creates a big
+    # problem if multiple copies are floating around in both the __init__
+    # function, and in the __dataclass_fields__ structure.
+    # Therefore, create module-level constants holding these values and
+    # pass those around instead
+    #
+    # If possible we use the `Field` class defined in the standard library
+    # module so that the information stored here is as close to a regular
+    # dataclass as is possible.
+    variables_assignment_stats = []
+    for name, field in fields.items():
+        if field.private:
+            continue  # doesn't appear in the public interface
+        for attrname in [ "default", "default_factory" ]:
+            f_def = getattr(field, attrname)
+            if f_def is MISSING or f_def.is_literal or f_def.is_name:
+                # some simple cases where we don't need to set up
+                # the variable as a module-level constant
+                continue
+            global_scope = node.scope.global_scope()
+            module_field_name = global_scope.mangle(
+                global_scope.mangle(Naming.dataclass_field_default_cname, node.class_name),
+                name)
+            # create an entry in the global scope for this variable to live
+            nn = ExprNodes.NameNode(f_def.pos, name=EncodedString(module_field_name))
+            nn.entry = global_scope.declare_var(nn.name, type=f_def.type or PyrexTypes.unspecified_type,
+                                                pos=f_def.pos, cname=nn.name, is_cdef=1)
+            # replace the field so that future users just receive the namenode
+            setattr(field, attrname, nn)
+
+            variables_assignment_stats.append(
+                Nodes.SingleAssignmentNode(f_def.pos,
+                                           lhs = nn,
+                                           rhs = f_def))
+
+    placeholders = {}
+    field_func = ExprNodes.AttributeNode(node.pos, obj = dataclass_module,
+                                         attribute=EncodedString("field"))
+    dc_fields = ExprNodes.DictNode(node.pos, key_value_pairs=[])
+    dc_fields_namevalue_assignments = []
+    for name, field in fields.items():
```

Comment:
Is there a reason that we loop through twice? I assume that it is because the second loop depends on the first loop having done the work for each item, or that it's simply for sectioning?